### PR TITLE
Login & Registration Panels + User Page Prototype

### PR DESCRIPTION
**Pull request by WawrzenskiPatryk**
### Worth to mention changes:
- authorization and authentication:
  - `LoginPage` & `RegisterPage` (no user input validation)
  - `UserAccountPage` prototype
  - logic for retrieving, storing and removing authentication token from API (register / login & logout logic)
  - error handling on 401 & 422 errors
  - router authentication guards
- environment & navigation based page title with `vue-meta`
- separation of production & staging environment
- `button-states` mixin and styling of `disabled` state for `BaseButton` component

### CODE REVIEW

File: src/components/user/UserDetailsForm.vue
Comment (WawrzenskiPatryk) [inline src/components/user/UserDetailsForm.vue]: That's a redundant leftover and will be deleted. Sorry about that.

File: src/pages/public/movies/MovieDetailPage.vue
Comment (szymonlicau) [inline src/pages/public/movies/MovieDetailPage.vue:73]: I'd say leave this for now until we touch upon composition API.
You are mixing composition API with options API and hence all the issues with this composable 🙏

File: src/router/authGuard.js
Comment (szymonlicau) [inline src/router/authGuard.js]: Might as well return it immediately
```suggestion
  return routes.reduce((routeNames, route) => {
    if (route.name) routeNames.push(route.name);
    return routeNames;
  }, []);
```

File: src/router/authGuard.js
Comment (szymonlicau) [inline src/router/authGuard.js:15]: It's much better to actually use a meta field on your routes for things like these, as the names can still change (or better yet, you might add some routes in different places which sometimes do/do not require authentication).
https://router.vuejs.org/guide/advanced/meta.html#route-meta-fields

File: src/services/api/auth.js
Comment (szymonlicau) [inline src/services/api/auth.js]: In all of these you can just return the result immediately, no need for an extra assignment.

File: src/services/api/data.js
Comment (szymonlicau) [inline src/services/api/data.js]: Same here 😄

File: src/services/api/data.js
Comment (szymonlicau) [inline src/services/api/data.js]: Just leave it to axios to handle this and **leave the default values as undefined** 😎:
```suggestion
  const screeningsResponse = await defaultClient.get(exactEndpoint, {
    params: {
      movie_id: movieId,
      date: date
    }
  });
```
This way any undefined value will not be added to query and there is no need to have all this logic here.

File: src/components/base/BaseInput.vue
Comment (szymonlicau) [inline src/components/base/BaseInput.vue:75]: While all of this seems to work, keep in mind that all of this extra logic dealing with either password input types or date input types should -ideally- be outside of this component.
In the end this is the **Base**Input. You can create specific components (`PasswordInput`, `DateInput`) to handle specific cases which would use the BaseInput undernearth and properly configure it/expand upon it.

**Review by szymonlicau — CHANGES_REQUESTED** [review]

File: src/router/authGuard.js
Comment (WawrzenskiPatryk) [reply src/router/authGuard.js]: I'm always in that little moments of returning the only value being processed in a function not so sure which way would be the better one - "const-ing" it or not 😅

I'm ok with either of two so I'll go with your suggestion 👍

File: src/services/api/data.js
Comment (WawrzenskiPatryk) [reply src/services/api/data.js]: Seems good that I've changed a little bit folder structure in `/services`, because all of that code have passed previous PRs and otherwise I would miss such an important tip, thank you!

File: src/components/base/BaseInput.vue
Comment (WawrzenskiPatryk) [reply src/components/base/BaseInput.vue:75]: Makes sense. Someone who didn't create the component might want to create himself e.g. `PasswordInput`, thinking that there couldn't be any password related code in "BASE-something" component. I can see that what I did might create a risk of redundancies in code and is just not that readable as it could be. Am I right?

Comment (szymonlicau) [reply src/components/base/BaseInput.vue:75]: Pretty much!
This will also make the component much more focused and be more "single-purpose" (remember the F.I.R.S.T. rules? 😄) which in turn - makes it more clean!
And if the password input would have let's imagine some more complex features - like I don't know in-site password-manager like thing - we would only extend the `PasswordInput` instead of building upon the ever-growing `BaseInput` 😄

File: src/services/api/data.js
Comment (WawrzenskiPatryk) [reply src/services/api/data.js]: In this scenario actually it causes a lot of errors:
```
export async function getAllMoviesData() {
  return await defaultClient.get(endpoint.MOVIES)?.data;
}
```
But something like this works fine if you would like it more:
```
export async function getAllMoviesData() {
  const { data } = await defaultClient.get(endpoint.MOVIES);
  return data;
}
```

And most of my logic in many other places is based on that those functions return the data property, hence the naming convention "getSomethingData"

Comment (szymonlicau) [reply src/services/api/data.js]: Ah right you return the data here!
The destructuring approach is absolutely fine then, my mistake, I didn't catch that 😅

Comment (WawrzenskiPatryk) [reply src/services/api/data.js]: I can also go with something like this since the code would repeat
```
async function getData(endpoint) {
  const { data } = await defaultClient.get(endpoint);
  return data;
}

export async function getAllMoviesData() {
  return getData(endpoint.MOVIES);
}

export async function getMovieData(movieId) {
  return getData(`${endpoint.MOVIES}/${movieId}`);
}

// ...
```

Comment (szymonlicau) [reply src/services/api/data.js]: Nah, there is no need unless you really want to - this is a bit overkill 😄

Comment (WawrzenskiPatryk) [reply src/services/api/data.js]: The whole file would be as short as this right now:

```
import defaultClient from '@/services/api/defaultClient.js';
import endpoint from '@/services/api/endpoint.js';

const getData = async (endpoint, params = {}) => {
  const { data } = await defaultClient.get(endpoint, params);
  return data;
};

export const getAllMoviesData = () => getData(endpoint.MOVIES);

export const getMovieData = movieId => getData(`${endpoint.MOVIES}/${movieId}`);

export const getMovieGenresData = () => getData(endpoint.GENRES);

export const getUserData = () => getData(endpoint.USER);

export const getSeancesData = (movieId = null, date = null) =>
  getData(endpoint.SEANCES, {
    params: {
      movie_id: movieId,
      date: date,
    },
  });
```

File: src/components/base/BaseInput.vue
Comment (WawrzenskiPatryk) [reply src/components/base/BaseInput.vue:75]: And would it be ok, if for now I leave it as such? I'd like to because I've made some changes on `develop` branch in the component and I'm afraid of extra conflicts. Will implement immediately after PR merge.

Comment (szymonlicau) [reply src/components/base/BaseInput.vue:75]: Sure! You can leave it for next PR 😄

File: src/services/api/data.js
Comment (WawrzenskiPatryk) [reply src/services/api/data.js]: Sorry I did not see any of your further comments, page did not update for me in real time 😅

Comment (WawrzenskiPatryk) [reply src/services/api/data.js]: And yes, overkill might be, yet I did it and I like it so let's leave it like that anyway 😅 Sorry again for missing your answers!

File: src/router/authGuard.js
Comment (WawrzenskiPatryk) [reply src/router/authGuard.js:15]: I'll check that and include in the next PR because it seems like I might need to take a moment to tweak the logic

**Review by szymonlicau — APPROVED** [review]
Looking good! Great job!
It would be reaaaaaally helpful if you would try to make the PRs a liiiiiiittle bit smaller 😄

**Pull request merged**
